src/cdk/layout/breakpoints-observer.ts

```diff
@@ -0,0 +1,96 @@
+
+/** The current state of a layout breakpoint. */
+export interface BreakpointState {
+  query: string;
+  matches: boolean;
+}
+
+interface Query {
+  observable: Observable<BreakpointState>;
+  mql: MediaQueryList;
+}
+
+@Injectable()
+export class BreakpointObserver implements OnDestroy {
+  /**  A map of all media queries currently being listened for. */
+  private _queries: Map<string, Query> = new Map();
+  /** A subject for all other observables to takeUntil based on. */
+  private _destroySubject: Subject<{}> = new Subject();
+
+  constructor(private mediaMatcher: MediaMatcher, private zone: NgZone) {}
+
+  /** Completes the active subject, signalling to all other observables to complete. */
+  ngOnDestroy() {
+    this._destroySubject.next();
+    this._destroySubject.complete();
+  }
+
+  /** Whether the query currently is matched. */
+  isMatched(value: string | string[]): boolean {
+    let queries = coerceArray(value);
+    return queries.some(mediaQuery => this._registerQuery(mediaQuery).mql.matches);
+  }
+
+  /**
+   * Gets an observable of results for the given queries that will emit new results for any changes
+   * in matching of the given queries.
+   */
+  observe(value: string | string[]): Observable<BreakpointState> {
+    let queries = coerceArray(value);
+    let observables = queries.map(query => this._registerQuery(query).observable);
+
+    return combineLatest(observables, (a: BreakpointState, b: BreakpointState) => {
+      return {
+        matches: !!((a && a.matches) || (b && b.matches)),
+        query: queries.join(', ')
+      };
+    });
+  }
+
+  /** Registers a specific query to be listened for. */
+  private _registerQuery(query: string): Query {
+    // Only set up a new MediaQueryList if it is not already being listened for.
+    if (this._queries.has(query)) {
+      return this._queries.get(query)!;
+    }
+
+    let mql: MediaQueryList = this.mediaMatcher.matchMedia(query);
+    // Create callback for match changes and add it is as a listener.
+    let queryObservable = RxChain.from(fromEventPattern(
+      // Listener callback methods are wrapped to be placed back in ngZone. Callbacks must be placed
+      // back into the zone because matchMedia is only included in Zone.js by loading the
+      // webapis-media-query.js file alongside the zone.js file.  Additionally, some browsers do not
+      // have MediaQueryList inherit from EventTarget, which causes inconsistencies in how Zone.js
+      // patches it.
+      (listener: MediaQueryListListener) => {
+        mql.addListener((e: MediaQueryList) => this.zone.run(() => listener(e)));
+      },
+      (listener: MediaQueryListListener) => {
+        mql.removeListener((e: MediaQueryList) => this.zone.run(() => listener(e)));
+      }))
+      .call(takeUntil, this._destroySubject)
+      .call(startWith, mql)
+      .call(map, (nextMql: MediaQueryList) => { return {query: query, matches: nextMql.matches}; })
```

Comment:
Don't need the braces and the `return`, just parens around the object literal
```ts
.call(map, (nextMql: MediaQueryList) => ({query: query, matches: nextMql.matches}))
```
Also, is the explicit `MediaQueryList` type necessary here? I think it should be inferred.

Reply:
We need to declare the `MediaQueryList` type, because the `fromEventPattern` observable emits an `any` type unless you use the `selector` method to transform what is returned, but we would just have to define the type in this method for typescript to know its type.  So I just went with declaring the type in the `map` parameter.